### DISCUSSION TITLE
Proper way of configuring cross compilation (Xbox Scarlett)

### DISCUSSION THREAD

**Discussion by Bvsemmer** [microsoft/vcpkg#38956] [Q&A] [answered]
I am preparing a dummy project configuration to try out vcpkg. I have a simple app and two C++ projects that I succesfully configured using vcpkg using Ninja and MSBuild on Windows. 

Since the intention is to use vcpkg as a proper package manager in our company, a second step I wanted to tryout is the cross-compilation functionality. I fail to get it working, and I feel I am missing information. 
I tried digging through the vcpkg scripts and toolchain files, but I need some extra help.

The intention is to get my application and the two dependent libraries compiled for Xbox Scarlett, since I do see support for that in the toolchains and in the supplied community scripts as well.

I have modified my `CMakePresets.json`, by adding an 'xbox' preset:

```
{
  "version": 2,
  "configurePresets": [
    {
      "name": "default",
      "generator": "Visual Studio 17 2022",
      "binaryDir": "${sourceDir}/build",
      "cacheVariables": {
        "CMAKE_TOOLCHAIN_FILE": "$env{VCPKG_ROOT}/scripts/buildsystems/vcpkg.cmake"
      }
    },
    {
      "name": "xbox",
      "generator": "Visual Studio 17 2022",
      "binaryDir": "${sourceDir}/build",
      "cacheVariables": {
        "CMAKE_TOOLCHAIN_FILE": "$env{VCPKG_ROOT}/scripts/buildsystems/vcpkg.cmake",
        "VCPKG_CHAINLOAD_TOOLCHAIN_FILE": "$env{VCPKG_ROOT}/scripts/toolchains/xbox.cmake"
      }
    }
  ]
}
```

And I run the build by calling `cmake --preset=xbox` in my project folder.

When I run this I am missing some important CMake variables, like:
VCPKG_TARGET_ARCHITECTURE, VCPKG_CRT_LINKAGE, VCPKG_LIBRARY_LINKAGE, VCPKG_XBOX_CONSOLE_TARGET

These are defined inside a triplet file supplied by the community `$env{VCPKG_ROOT}/triplets/community/x64-xbox-scarlett.cmake`. But I do not see how I can include that file during the build step for example.  `vcpkg.cmake` is not setup to include that before it actually includes `VCPKG_CHAINLOAD_TOOLCHAIN_FILE`. 

Clearly I am missing something here. What is the correct way of generating my project with a different toolchain?

Thanks a lot in advance!

**Comment by Neumann-A** [comment]
you are expected to have your own toolchain in VCPKG_CHAINLOAD_TOOLCHAIN_FILE not one from vcpkg. The one from vcpkg is for internal use. If you want to use it externally you have to define all variables in your preset which this internal toolchain expects.

**Reply by Bvsemmer** [reply]
Ok, I guess that makes sense. So the combination in the 'xbox' preset with a `CMAKE_TOOLCHAIN_FILE` and `VCPKG_CHAINLOAD_TOOLCHAIN_FILE` is the correct way of approaching this? 

Thank you for your reply!

**Reply by Neumann-A** [reply]
In general yes. It is always the correct approach if you want to use vcpkg together with your own toolchain file.

**Reply by Bvsemmer** [reply]
thanks

**Comment by Bvsemmer** [comment]
I have my own toolchain file now for Scarlett and "another console platform" that is using Clang as it's compiler and linker. For the main application, things are looking OK now. The project is built using the correct toolchain.  
But it looks like the _dependencies_ are not built with the correct triplet setting, even though it seems to hint at it during the output.

I have a single test Application with two libraries as dependencies (**bvslib** and **tinyxml2** both story in a custom local registry)

This is what I get in the output:
![image](https://github.com/microsoft/vcpkg/assets/170676/0b0b3920-c042-4f4d-a326-b8dc2d9bd457)

The output libraries do not contain LLVM bitcode though. When I build this in Visual Studio, I get the following output:
```
prospero-lld : warning : vcpkg_installed\x64-playstation-prospero\debug\lib\tinyxml2.lib: archive member 'CMakeFiles\tinyxml2.dir\tinyxml2.cpp.obj' is neither ET_REL nor LLVM bitcode
2>prospero-lld : warning : vcpkg_installed\x64-playstation-prospero\debug\lib\bvslib.lib: archive member 'CMakeFiles\bvslib.dir\bvslib.cpp.obj' is neither ET_REL nor LLVM bitcode
```

So seems to me the libraries are still built with the Windows toolchain?

I have two questions related to that:
- Is there a way to only build dependencies for a project and install in vcpkg for a specific triplet? 
- How do I best go about trying to debug this? 

Thank you!

**Reply by Neumann-A** [reply]
Check the install/build logs if the correct compiler/linker is used with the correct flags?
Sometimes project override flags, this needs to be patched out.

**Comment by dg0yt** [comment]
AFAICT you must set `VCPKG_TARGET_TRIPLET`.
https://learn.microsoft.com/en-us/vcpkg/users/buildsystems/cmake-integration#vcpkg_target_triplet

**Comment by Bvsemmer** [comment, marked as answer]
Hey all,

Thank you for your replies! After some digging in the vcpkg-tool code and reading the vcpkg cmake file, I found the problem I was having. 
I had two issues:
- vcpkg.exe does not pass along the Environment Variables by default. I had to specifiy some using VCPKG_KEEP_ENV_VARS to be passed to the CMake environment called by vcpkg. 
- I had to respecify VCPKG_CHAINLOAD_TOOLCHAIN_FILE inside my triplet file. Which is a bit of a pitty, cause I have to specify it inside the triplet file and the CMakePresets.json.
